### PR TITLE
feat(sm-integration): Add SageMaker training as a backend.

### PR DESCRIPTION
**Pull request by Linbo-Liu**
*Issue #, if available:*

*Description of changes:*
## What

Adds an experimental **SageMaker Training Sessions** backend: a GRPO loop that trains an AgentCore Runtime (ACR)-deployed agent with **no local GPU cluster**. SageMaker hosts the policy weights, the sampler, and the optimizer behind an SDK, so the RL loop itself is a plain single-process asyncio program that can run on a laptop or a small EC2 box. Token capture reuses the in-repo rollout gateway, exactly as the experimental verl backend does.

## How it fits together

The only new engine seam is `SageMakerSdkBackend` (`rollout_gateway/sampling_backends/sagemaker_sdk.py`) — a gateway `SamplingBackend` over the SageMaker `SamplingClient` that maps `token_ids -> token_ids + logprobs`. Like `TinkerSdkBackend`, it does not render: the gateway owns tokenization, which keeps loss-masking well-defined and matches the existing placement rule (independently reachable hosted SDK → `sampling_backends/`).

## The training loop

`backends/experimental/sagemaker/` is driven by a single YAML config (`config.py` / `config.yaml.example`):

| File | Role |
|---|---|
| `train_grpo.py` | GRPO loop: rollout → advantages → `forward_backward` → `optim_step` → rebind sampler |
| `rollout.py` | One ACR rollout: create gateway session, invoke the agent, await the S3 result, drain trajectories |
| `datum.py` | `TraceRecord` + advantage → SageMaker training datum |
| `config.py` / `config.yaml.example` | Typed config dataclass and a template to copy |
| `prepare_datasets/prepare_gsm8k.py` | GSM8K → payload-column Parquet |

## Tests

Tested on GSM8k. Stable training for 200+ steps. Test accuracy increases from ~70% -> ~90% with `gpt-oss-20b`.

## Docs

New `docs/site/src/content/docs/guides/sagemaker-backend-setup.md` (installation, config reference, what the loop does step by step, current limits), wired into the sidebar and linked from `index.mdx` and `guides/overview.mdx` alongside slime / rllm / verl. A short `SETUP.md` sits next to the code for readers who get there from the source tree.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/agentcore_rl_toolkit/backends/experimental/sagemaker/datum.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    if prompt_length < 1:
+        return None
+
+    input_tokens = token_ids[:-1]
+    target_tokens = token_ids[1:]
+    seq_len = len(input_tokens)
+
+    weights = [0.0] * (prompt_length - 1) + [float(m) for m in record.loss_mask]
+    logprobs = [0.0] * (prompt_length - 1) + list(record.logprobs)
+    advantages = [0.0] * (prompt_length - 1) + [float(advantage)] * response_length
```
Comment (luyuzhe111) [inline src/agentcore_rl_toolkit/backends/experimental/sagemaker/datum.py]: Just a clarification question: do all three supported losses honor weights as a loss mask? 

Tool/context tokens receive nonzero advantages here; if weights is ignored, their advantages need to be zeroed using record.loss_mask.

File: src/agentcore_rl_toolkit/backends/experimental/sagemaker/SETUP.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+uv pip install sagemaker-train
+uv pip install -e ".[gateway]"
+uv pip install transformers==5.12
+```
+
+## 2. Prepare training dataset
+Scripts for preparing AgentCore-compatible dataset are in `src/agentcore_rl_toolkit/backends/experimental/sagemaker/prepare_datasets`. For example, to prepare a `gsm8k` dataset, run:
+```bash
+python src/agentcore_rl_toolkit/backends/experimental/verl/examples/math_agent/preprocess_gsm8k.py \
```
Comment (luyuzhe111) [inline src/agentcore_rl_toolkit/backends/experimental/sagemaker/SETUP.md]: The gateway-based verl backend graduated in #116, so this script path no longer exists. can use the new sagemaker preparation script?

would be great if you can clean up other stale references to experimental verl too.

File: docs/site/src/content/docs/guides/sagemaker-backend-setup.md [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+description: Train an AgentCore Runtime-deployed agent with Amazon SageMaker Training Sessions — no local GPUs required.
+---
+
+This doc describes how to train an AgentCore Runtime-deployed agent with
+**Amazon SageMaker Training Sessions**. Unlike the
+[slime](/agentcore-rl-toolkit/guides/slime-backend-setup/) /
+[verl](/agentcore-rl-toolkit/guides/verl-backend-setup/) backends, there is
+**no GPU cluster required for the training**: SageMaker hosts the policy weights,
+the sampler, and the optimizer behind an SDK, and the RL loop itself runs as a
+plain Python process on your laptop or a small EC2 box.
```
Comment (luyuzhe111) [inline docs/site/src/content/docs/guides/sagemaker-backend-setup.md]: Could we document the current networking requirement here? The ACR runtime needs VPC mode in the same VPC as the training-loop host, with access to the gateway port. Running directly from a laptop isn't supported by this setup?

**Review by luyuzhe111 — CHANGES_REQUESTED** [review]

File: src/agentcore_rl_toolkit/backends/experimental/sagemaker/datum.py (hunk 1 above)
Comment (Linbo-Liu) [reply src/agentcore_rl_toolkit/backends/experimental/sagemaker/datum.py]: Good point. I thought SM training client would multiply weights and advantage. But better to double guard it.

File: src/agentcore_rl_toolkit/backends/experimental/sagemaker/SETUP.md (hunk 2 above)
Comment (Linbo-Liu) [reply src/agentcore_rl_toolkit/backends/experimental/sagemaker/SETUP.md]: Thanks for the careful review. Removed the stale references

File: docs/site/src/content/docs/guides/sagemaker-backend-setup.md (hunk 3 above)
Comment (Linbo-Liu) [reply docs/site/src/content/docs/guides/sagemaker-backend-setup.md]: Makes sense.

**Review by luyuzhe111 — APPROVED** [review]
